// @twurple/eventsub/lib/events/EventSubChannelHypeTrainEndEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSubChannelHypeTrainEndEvent = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * An EventSub event representing the end of a Hype train event.
 */
let EventSubChannelHypeTrainEndEvent = class EventSubChannelHypeTrainEndEvent {
    /** @private */
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this._data.broadcaster_user_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._data.broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this._data.broadcaster_user_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.broadcaster_user_id));
    }
    /**
     * The level the Hype Train ended on.
     */
    get level() {
        return this._data.level;
    }
    /**
     * The total points contributed to the Hype Train.
     */
    get total() {
        return this._data.total;
    }
    /**
     * The contributors with the most points, for both bits and subscriptions.
     */
    get topContributions() {
        return this._data.top_contributions;
    }
    /**
     * The time when the Hype Train started.
     */
    get startDate() {
        return new Date(this._data.started_at);
    }
    /**
     * The time when the Hype Train ended.
     */
    get endDate() {
        return new Date(this._data.ended_at);
    }
    /**
     * The time when the Hype Train cooldown ends.
     */
    get cooldownEndDate() {
        return new Date(this._data.cooldown_ends_at);
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], EventSubChannelHypeTrainEndEvent.prototype, "_client", void 0);
EventSubChannelHypeTrainEndEvent = tslib_1.__decorate([
    common_1.rtfm('eventsub', 'EventSubChannelHypeTrainEndEvent', 'broadcasterId')
], EventSubChannelHypeTrainEndEvent);
exports.EventSubChannelHypeTrainEndEvent = EventSubChannelHypeTrainEndEvent;
