// @twurple/api/es/api/helix/channel/HelixChannel.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * A Twitch channel.
 */
let HelixChannel = class HelixChannel {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The ID of the channel.
     */
    get id() {
        return this._data.broadcaster_id;
    }
    /**
     * The name of the channel.
     */
    get name() {
        return this._data.broadcaster_login;
    }
    /**
     * The display name of the channel.
     */
    get displayName() {
        return this._data.broadcaster_name;
    }
    /**
     * The language of the channel.
     */
    get language() {
        return this._data.broadcaster_language;
    }
    /**
     * The ID of the game currently played on the channel.
     */
    get gameId() {
        return this._data.game_id;
    }
    /**
     * The name of the game currently played on the channel.
     */
    get gameName() {
        return this._data.game_name;
    }
    /**
     * Retrieves information about the game that is being played on the stream.
     */
    async getGame() {
        return (await this._client.helix.games.getGameById(this._data.game_id));
    }
    /**
     * The title of the channel.
     */
    get title() {
        return this._data.title;
    }
};
__decorate([
    Enumerable(false)
], HelixChannel.prototype, "_data", void 0);
__decorate([
    Enumerable(false)
], HelixChannel.prototype, "_client", void 0);
HelixChannel = __decorate([
    rtfm('api', 'HelixChannel', 'id')
], HelixChannel);
export { HelixChannel };
