// @twurple/api/lib/api/helix/webHooks/HelixWebHookSubscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixWebHookSubscription = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A subscription to a Twitch WebHook.
 */
let HelixWebHookSubscription = class HelixWebHookSubscription {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The topic the WebHook is listening to.
     */
    get topicUrl() {
        return this._data.topic;
    }
    /**
     * The URL that will be called for every subscribed event.
     */
    get callbackUrl() {
        return this._data.callback;
    }
    /**
     * The time when the subscription will expire.
     */
    get expiryDate() {
        return new Date(this._data.expires_at);
    }
    /**
     * Unsubscribe from the WebHook.
     */
    async unsubscribe() {
        await this._client.helix.webHooks.sendHubRequest({
            mode: 'unsubscribe',
            topicUrl: this.topicUrl,
            callbackUrl: this.callbackUrl
        });
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixWebHookSubscription.prototype, "_data", void 0);
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixWebHookSubscription.prototype, "_client", void 0);
HelixWebHookSubscription = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixWebHookSubscription')
], HelixWebHookSubscription);
exports.HelixWebHookSubscription = HelixWebHookSubscription;
