// @twurple/api/lib/api/helix/channelPoints/HelixCustomRewardRedemption.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixCustomRewardRedemption = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A redemption of a custom Channel Points reward.
 */
let HelixCustomRewardRedemption = class HelixCustomRewardRedemption {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The ID of the redemption.
     */
    get id() {
        return this._data.id;
    }
    /**
     * The ID of the broadcaster where the reward was redeemed.
     */
    get broadcasterId() {
        return this._data.broadcaster_id;
    }
    /**
     * The name of the broadcaster where the reward was redeemed.
     */
    get broadcasterName() {
        return this._data.broadcaster_login;
    }
    /**
     * The display name of the broadcaster where the reward was redeemed.
     */
    get broadcasterDisplayName() {
        return this._data.broadcaster_name;
    }
    /**
     * Retrieves more information about the broadcaster where the reward was redeemed.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.broadcaster_id));
    }
    /**
     * The ID of the user that redeemed the reward.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the user that redeemed the reward.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the user that redeemed the reward.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * Retrieves more information about the user that redeemed the reward.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
    /**
     * The text the user wrote when redeeming the reward.
     */
    get userInput() {
        return this._data.user_input;
    }
    /**
     * Whether the redemption was fulfilled.
     */
    get isFulfilled() {
        return this._data.status === 'FULFILLED';
    }
    /**
     * Whether the redemption was canceled.
     */
    get isCanceled() {
        return this._data.status === 'CANCELED';
    }
    /**
     * The date and time when the reward was redeemed.
     */
    get redemptionDate() {
        return new Date(this._data.redeemed_at);
    }
    /**
     * The ID of the reward that was redeemed.
     */
    get rewardId() {
        return this._data.reward.id;
    }
    /**
     * Retrieves more info about the reward that was redeemed.
     */
    async getReward() {
        return (await this._client.helix.channelPoints.getCustomRewardById(this._data.broadcaster_id, this._data.reward.id));
    }
    /**
     * The title of the reward that was redeemed.
     */
    get rewardTitle() {
        return this._data.reward.title;
    }
    /**
     * The prompt of the reward that was redeemed.
     */
    get rewardPrompt() {
        return this._data.reward.prompt;
    }
    /**
     * The cost of the reward that was redeemed.
     */
    get rewardCost() {
        return this._data.reward.cost;
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixCustomRewardRedemption.prototype, "_data", void 0);
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixCustomRewardRedemption.prototype, "_client", void 0);
HelixCustomRewardRedemption = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixCustomRewardRedemption', 'id')
], HelixCustomRewardRedemption);
exports.HelixCustomRewardRedemption = HelixCustomRewardRedemption;
