// @twurple/api/lib/api/helix/subscriptions/HelixSubscriptionEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixSubscriptionEvent = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
const HelixSubscription_1 = require("./HelixSubscription");
/**
 * An event that indicates the change of a subscription status, i.e. subscribing, unsubscribing or sending the monthly notification.
 */
let HelixSubscriptionEvent = class HelixSubscriptionEvent extends HelixSubscription_1.HelixSubscription {
    /** @private */
    constructor(eventData, client) {
        super(eventData.event_data, client);
        this._eventData = eventData;
    }
    /**
     * The unique ID of the subscription event.
     */
    get eventId() {
        return this._eventData.id;
    }
    /**
     * The type of the subscription event.
     */
    get eventType() {
        return this._eventData.event_type;
    }
    /**
     * The date of the subscription event.
     */
    get eventDate() {
        return new Date(this._eventData.event_timestamp);
    }
    /**
     * The version of the subscription event.
     */
    get eventVersion() {
        return this._eventData.version;
    }
    /**
     * The message sent with the subscription event.
     */
    get eventMessage() {
        var _a;
        return (_a = this._eventData.event_data.message) !== null && _a !== void 0 ? _a : '';
    }
};
HelixSubscriptionEvent = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixSubscriptionEvent', 'userId')
], HelixSubscriptionEvent);
exports.HelixSubscriptionEvent = HelixSubscriptionEvent;
