// @twurple/api/lib/api/helix/subscriptions/HelixSubscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixSubscription = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A (paid) subscription of a user to a broadcaster.
 */
let HelixSubscription = class HelixSubscription {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The user ID of the broadcaster.
     */
    get broadcasterId() {
        return this._data.broadcaster_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._data.broadcaster_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this._data.broadcaster_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.broadcaster_id));
    }
    /**
     * The user ID of the gifter.
     */
    get gifterId() {
        return this._data.gifter_id;
    }
    /**
     * The name of the gifter.
     */
    get gifterName() {
        return this._data.gifter_login;
    }
    /**
     * The display name of the gifter.
     */
    get gifterDisplayName() {
        return this._data.gifter_name;
    }
    /**
     * Retrieves more information about the gifter.
     */
    async getGifter() {
        return (await this._client.helix.users.getUserById(this._data.gifter_id));
    }
    /**
     * Whether the subscription has been gifted by another user.
     */
    get isGift() {
        return this._data.is_gift;
    }
    /**
     * The tier of the subscription.
     */
    get tier() {
        return this._data.tier;
    }
    /**
     * The user ID of the subscribed user.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the subscribed user.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the subscribed user.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * Retrieves more information about the subscribed user.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixSubscription.prototype, "_data", void 0);
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixSubscription.prototype, "_client", void 0);
HelixSubscription = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixSubscription', 'userId')
], HelixSubscription);
exports.HelixSubscription = HelixSubscription;
