// @twurple/api/lib/api/kraken/KrakenApiGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KrakenApiGroup = void 0;
const tslib_1 = require("tslib");
const cache_decorators_1 = require("@d-fischer/cache-decorators");
const common_1 = require("@twurple/common");
const BaseApi_1 = require("../BaseApi");
const BitsApi_1 = require("./bits/BitsApi");
const ChannelApi_1 = require("./channel/ChannelApi");
const ChatApi_1 = require("./chat/ChatApi");
const SearchApi_1 = require("./search/SearchApi");
const StreamApi_1 = require("./stream/StreamApi");
const TeamApi_1 = require("./team/TeamApi");
const UserApi_1 = require("./user/UserApi");
const VideoApi_1 = require("./video/VideoApi");
/**
 * Groups all API calls available in Kraken v5.
 *
 * Can be accessed using {@ApiClient#kraken}.
 */
let KrakenApiGroup = class KrakenApiGroup extends BaseApi_1.BaseApi {
    /**
     * The API methods that deal with bits.
     */
    get bits() {
        return new BitsApi_1.BitsApi(this._client);
    }
    /**
     * The API methods that deal with channels.
     */
    get channels() {
        return new ChannelApi_1.ChannelApi(this._client);
    }
    /**
     * The API methods that deal with chat.
     */
    get chat() {
        return new ChatApi_1.ChatApi(this._client);
    }
    /**
     * The API methods that deal with searching.
     */
    get search() {
        return new SearchApi_1.SearchApi(this._client);
    }
    /**
     * The API methods that deal with streams.
     */
    get streams() {
        return new StreamApi_1.StreamApi(this._client);
    }
    /**
     * The API methods that deal with teams.
     */
    get teams() {
        return new TeamApi_1.TeamApi(this._client);
    }
    /**
     * The API methods that deal with videos.
     */
    get videos() {
        return new VideoApi_1.VideoApi(this._client);
    }
    /**
     * The API methods that deal with users.
     */
    get users() {
        return new UserApi_1.UserApi(this._client);
    }
};
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "bits", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "channels", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "chat", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "search", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "streams", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "teams", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "videos", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], KrakenApiGroup.prototype, "users", null);
KrakenApiGroup = tslib_1.__decorate([
    cache_decorators_1.Cacheable,
    common_1.rtfm('api', 'KrakenApiGroup')
], KrakenApiGroup);
exports.KrakenApiGroup = KrakenApiGroup;
