// @twurple/api/lib/api/badges/ChatBadgeVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatBadgeVersion = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A version of a badge.
 */
let ChatBadgeVersion = class ChatBadgeVersion {
    /** @private */
    constructor(data) {
        this._data = data;
    }
    /**
     * The action to execute when the badge is clicked.
     */
    get clickAction() {
        return this._data.click_action;
    }
    /**
     * The URL to visit when the badge is clicked.
     *
     * Only applies if clickAction === 'visit_url'.
     */
    get clickUrl() {
        return this._data.click_url;
    }
    /**
     * The description of the badge.
     */
    get description() {
        return this._data.description;
    }
    /**
     * Gets an image URL for the given scale.
     *
     * @param scale The scale of the badge image.
     */
    getImageUrl(scale) {
        // eslint-disable-next-line @typescript-eslint/non-nullable-type-assertion-style
        return this._data[`image_url_${scale}x`];
    }
    /**
     * The title of the badge.
     */
    get title() {
        return this._data.title;
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], ChatBadgeVersion.prototype, "_data", void 0);
ChatBadgeVersion = tslib_1.__decorate([
    common_1.rtfm('api', 'ChatBadgeVersion')
], ChatBadgeVersion);
exports.ChatBadgeVersion = ChatBadgeVersion;
