// @twurple/api/es/api/helix/moderation/HelixModeratorEvent.mjs
import { __decorate } from "tslib";
import { rtfm } from '@twurple/common';
import { HelixModerator } from "./HelixModerator.mjs";
/**
 * An event that indicates the change of a moderator status, i.e. gaining or losing moderation privileges.
 */
let HelixModeratorEvent = class HelixModeratorEvent extends HelixModerator {
    /** @private */
    constructor(eventData, client) {
        super(eventData.event_data, client);
        this._eventData = eventData;
    }
    /**
     * The unique ID of the moderator event.
     */
    get eventId() {
        return this._eventData.id;
    }
    /**
     * The type of the moderator event.
     */
    get eventType() {
        return this._eventData.event_type;
    }
    /**
     * The date of the moderator event.
     */
    get eventDate() {
        return new Date(this._eventData.event_timestamp);
    }
    /**
     * The version of the moderator event.
     */
    get eventVersion() {
        return this._eventData.version;
    }
    /**
     * The id of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterId() {
        return this._eventData.event_data.broadcaster_id;
    }
    /**
     * The name of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterName() {
        return this._eventData.event_data.broadcaster_login;
    }
    /**
     * The display name of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterDisplayName() {
        return this._eventData.event_data.broadcaster_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._eventData.event_data.broadcaster_id));
    }
};
HelixModeratorEvent = __decorate([
    rtfm('api', 'HelixModeratorEvent', 'userId')
], HelixModeratorEvent);
export { HelixModeratorEvent };
