// @twurple/eventsub/es/events/EventSubChannelRedemptionUpdateEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * An EventSub event representing a Channel Points redemption being updated.
 */
let EventSubChannelRedemptionUpdateEvent = class EventSubChannelRedemptionUpdateEvent {
    /** @private */
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    /**
     * The ID of the redemption being updated.
     */
    get id() {
        return this._data.id;
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this._data.broadcaster_user_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._data.broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this._data.broadcaster_user_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.broadcaster_user_id));
    }
    /**
     * The ID of the user.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the user.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the user.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * Retrieves more information about the user.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
    /**
     * The input text given by the user.
     *
     * If there is no input to be given, this is an empty string.
     */
    get input() {
        return this._data.user_input;
    }
    /**
     * The status of the redemption.
     */
    get status() {
        return this._data.status;
    }
    /**
     * The ID of the reward that was redeemed.
     */
    get rewardId() {
        return this._data.reward.id;
    }
    /**
     * The title of the reward that was redeemed.
     */
    get rewardTitle() {
        return this._data.reward.title;
    }
    /**
     * The cost of the reward that was redeemed.
     */
    get rewardCost() {
        return this._data.reward.cost;
    }
    /**
     * The description of the reward that was redeemed.
     */
    get rewardPrompt() {
        return this._data.reward.prompt;
    }
    /**
     * The time when the user redeemed the reward.
     */
    get redemptionDate() {
        return new Date(this._data.redeemed_at);
    }
};
__decorate([
    Enumerable(false)
], EventSubChannelRedemptionUpdateEvent.prototype, "_client", void 0);
EventSubChannelRedemptionUpdateEvent = __decorate([
    rtfm('eventsub', 'EventSubChannelRedemptionUpdateEvent', 'id')
], EventSubChannelRedemptionUpdateEvent);
export { EventSubChannelRedemptionUpdateEvent };
