// @twurple/eventsub/es/events/EventSubChannelCheerEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * An EventSub event representing a user cheering bits.
 */
let EventSubChannelCheerEvent = class EventSubChannelCheerEvent {
    /** @private */
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    /**
     * The ID of the cheering user, null if anonymous.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the cheering user, null if anonymous.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the cheering user, null if anonymous.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * Retrieves more information about the user.
     */
    async getUser() {
        return this._data.user_id ? await this._client.helix.users.getUserById(this._data.user_id) : null;
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this._data.broadcaster_user_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._data.broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this._data.broadcaster_user_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.broadcaster_user_id));
    }
    /**
     * Whether the cheering user chose to be anonymous.
     */
    get isAnonymous() {
        return this._data.is_anonymous;
    }
    /**
     * The message sent with the cheer.
     */
    get message() {
        return this._data.message;
    }
    /**
     * The amount of bits cheered.
     */
    get bits() {
        return this._data.bits;
    }
};
__decorate([
    Enumerable(false)
], EventSubChannelCheerEvent.prototype, "_client", void 0);
EventSubChannelCheerEvent = __decorate([
    rtfm('eventsub', 'EventSubChannelCheerEvent', 'userId')
], EventSubChannelCheerEvent);
export { EventSubChannelCheerEvent };
