// @twurple/eventsub/es/events/EventSubUserAuthorizationRevokeEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * An EventSub event representing a user revoking authorization for an application.
 */
let EventSubUserAuthorizationRevokeEvent = class EventSubUserAuthorizationRevokeEvent {
    /** @private */
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    /**
     * The ID of the user who revoked their authorization.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the user who revoked their authorization.
     *
     * This is `null` if the user no longer exists.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the user who revoked their authorization.
     *
     * This is `null` if the user no longer exists.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * Retrieves more information about the user.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
    /**
     * The Client ID of the application that the user revoked authorization for.
     */
    get clientId() {
        return this._data.client_id;
    }
};
__decorate([
    Enumerable(false)
], EventSubUserAuthorizationRevokeEvent.prototype, "_client", void 0);
EventSubUserAuthorizationRevokeEvent = __decorate([
    rtfm('eventsub', 'EventSubUserAuthorizationRevokeEvent', 'userId')
], EventSubUserAuthorizationRevokeEvent);
export { EventSubUserAuthorizationRevokeEvent };
