// @twurple/eventsub/es/events/EventSubChannelRaidEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * An EventSub event representing a broadcaster raiding another broadcaster.
 */
let EventSubChannelRaidEvent = class EventSubChannelRaidEvent {
    /** @private */
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    /**
     * The ID of the raiding broadcaster.
     */
    get raidingBroadcasterId() {
        return this._data.from_broadcaster_user_id;
    }
    /**
     * The name of the raiding broadcaster.
     */
    get raidingBroadcasterName() {
        return this._data.from_broadcaster_user_login;
    }
    /**
     * The display name of the raiding broadcaster.
     */
    get raidingBroadcasterDisplayName() {
        return this._data.from_broadcaster_user_name;
    }
    /**
     * Retrieves more information about the raiding broadcaster.
     */
    async getRaidingBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.from_broadcaster_user_id));
    }
    /**
     * The ID of the raided broadcaster.
     */
    get raidedBroadcasterId() {
        return this._data.to_broadcaster_user_id;
    }
    /**
     * The name of the raided broadcaster.
     */
    get raidedBroadcasterName() {
        return this._data.to_broadcaster_user_login;
    }
    /**
     * The display name of the raided broadcaster.
     */
    get raidedBroadcasterDisplayName() {
        return this._data.to_broadcaster_user_name;
    }
    /**
     * Retrieves more information about the raided broadcaster.
     */
    async getRaidedBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.to_broadcaster_user_id));
    }
    /**
     * The amount of viewers in the raid.
     */
    get viewers() {
        return this._data.viewers;
    }
};
__decorate([
    Enumerable(false)
], EventSubChannelRaidEvent.prototype, "_client", void 0);
EventSubChannelRaidEvent = __decorate([
    rtfm('eventsub', 'EventSubChannelRaidEvent', 'raidedBroadcasterId')
], EventSubChannelRaidEvent);
export { EventSubChannelRaidEvent };
