// @twurple/eventsub/es/events/EventSubStreamOnlineEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * An EventSub event representing a stream going live.
 */
let EventSubStreamOnlineEvent = class EventSubStreamOnlineEvent {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this._data.broadcaster_user_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._data.broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this._data.broadcaster_user_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.broadcaster_user_id));
    }
    /**
     * The type of the stream going live.
     */
    get streamType() {
        return this._data.type;
    }
    /**
     * The date and time when the stream was started.
     */
    get startDate() {
        return new Date(this._data.started_at);
    }
};
__decorate([
    Enumerable(false)
], EventSubStreamOnlineEvent.prototype, "_data", void 0);
__decorate([
    Enumerable(false)
], EventSubStreamOnlineEvent.prototype, "_client", void 0);
EventSubStreamOnlineEvent = __decorate([
    rtfm('eventsub', 'EventSubStreamOnlineEvent', 'broadcasterId')
], EventSubStreamOnlineEvent);
export { EventSubStreamOnlineEvent };
