// @twurple/eventsub/es/events/EventSubChannelSubscriptionEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * An EventSub event representing a channel subscription.
 */
let EventSubChannelSubscriptionEvent = class EventSubChannelSubscriptionEvent {
    /** @private */
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    /**
     * The ID of the subscribing user.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the subscribing user.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the subscribing user.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * Retrieves more information about the subscription.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this._data.broadcaster_user_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._data.broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this._data.broadcaster_user_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._data.broadcaster_user_id));
    }
    /**
     * The tier of the subscription, either 1000, 2000 or 3000.
     */
    get tier() {
        return this._data.tier;
    }
    /**
     * Whether the subscription has been gifted.
     */
    get isGift() {
        return this._data.is_gift;
    }
};
__decorate([
    Enumerable(false)
], EventSubChannelSubscriptionEvent.prototype, "_client", void 0);
EventSubChannelSubscriptionEvent = __decorate([
    rtfm('eventsub', 'EventSubChannelSubscriptionEvent', 'userId')
], EventSubChannelSubscriptionEvent);
export { EventSubChannelSubscriptionEvent };
