// @twurple/api/es/api/helix/bits/HelixBitsLeaderboardEntry.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { rtfm } from '@twurple/common';
/**
 * A Bits leaderboard entry.
 */
let HelixBitsLeaderboardEntry = class HelixBitsLeaderboardEntry {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The ID of the user on the leaderboard.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the user on the leaderboard.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the user on the leaderboard.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * The position of the user on the leaderboard.
     */
    get rank() {
        return this._data.rank;
    }
    /**
     * The amount of bits used in the given period of time.
     */
    get amount() {
        return this._data.score;
    }
    /**
     * Retrieves the user of entry on the leaderboard.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
};
__decorate([
    Enumerable(false)
], HelixBitsLeaderboardEntry.prototype, "_data", void 0);
__decorate([
    Enumerable(false)
], HelixBitsLeaderboardEntry.prototype, "_client", void 0);
HelixBitsLeaderboardEntry = __decorate([
    rtfm('api', 'HelixBitsLeaderboardEntry', 'userId')
], HelixBitsLeaderboardEntry);
export { HelixBitsLeaderboardEntry };
