// @twurple/api/es/api/kraken/KrakenApiGroup.mjs
import { __decorate } from "tslib";
import { Cacheable, CachedGetter } from '@d-fischer/cache-decorators';
import { rtfm } from '@twurple/common';
import { BaseApi } from "../BaseApi.mjs";
import { BitsApi } from "./bits/BitsApi.mjs";
import { ChannelApi } from "./channel/ChannelApi.mjs";
import { ChatApi } from "./chat/ChatApi.mjs";
import { SearchApi } from "./search/SearchApi.mjs";
import { StreamApi } from "./stream/StreamApi.mjs";
import { TeamApi } from "./team/TeamApi.mjs";
import { UserApi } from "./user/UserApi.mjs";
import { VideoApi } from "./video/VideoApi.mjs";
/**
 * Groups all API calls available in Kraken v5.
 *
 * Can be accessed using {@ApiClient#kraken}.
 */
let KrakenApiGroup = class KrakenApiGroup extends BaseApi {
    /**
     * The API methods that deal with bits.
     */
    get bits() {
        return new BitsApi(this._client);
    }
    /**
     * The API methods that deal with channels.
     */
    get channels() {
        return new ChannelApi(this._client);
    }
    /**
     * The API methods that deal with chat.
     */
    get chat() {
        return new ChatApi(this._client);
    }
    /**
     * The API methods that deal with searching.
     */
    get search() {
        return new SearchApi(this._client);
    }
    /**
     * The API methods that deal with streams.
     */
    get streams() {
        return new StreamApi(this._client);
    }
    /**
     * The API methods that deal with teams.
     */
    get teams() {
        return new TeamApi(this._client);
    }
    /**
     * The API methods that deal with videos.
     */
    get videos() {
        return new VideoApi(this._client);
    }
    /**
     * The API methods that deal with users.
     */
    get users() {
        return new UserApi(this._client);
    }
};
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "bits", null);
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "channels", null);
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "chat", null);
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "search", null);
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "streams", null);
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "teams", null);
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "videos", null);
__decorate([
    CachedGetter()
], KrakenApiGroup.prototype, "users", null);
KrakenApiGroup = __decorate([
    Cacheable,
    rtfm('api', 'KrakenApiGroup')
], KrakenApiGroup);
export { KrakenApiGroup };
