// @twurple/api/lib/ApiClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiClient = void 0;
const tslib_1 = require("tslib");
const cache_decorators_1 = require("@d-fischer/cache-decorators");
const api_call_1 = require("@twurple/api-call");
const auth_1 = require("@twurple/auth");
const common_1 = require("@twurple/common");
const BadgesApi_1 = require("./api/badges/BadgesApi");
const HelixApiGroup_1 = require("./api/helix/HelixApiGroup");
const HelixRateLimiter_1 = require("./api/helix/HelixRateLimiter");
const KrakenApiGroup_1 = require("./api/kraken/KrakenApiGroup");
const UnsupportedApi_1 = require("./api/unsupported/UnsupportedApi");
const ConfigError_1 = require("./Errors/ConfigError");
/**
 * An API client for the Twitch Kraken and Helix APIs.
 */
let ApiClient = class ApiClient {
    /**
     * Creates a new API client instance.
     *
     * @param config Configuration for the client instance.
     */
    constructor(config) {
        if (!config.authProvider) {
            throw new ConfigError_1.ConfigError('No auth provider given. Please supply the `authProvider` option.');
        }
        this._helixRateLimiter = new HelixRateLimiter_1.HelixRateLimiter({ logger: config.logger });
        this._config = config;
    }
    /**
     * Requests scopes from the auth provider.
     *
     * @param scopes The scopes to request.
     */
    async requestScopes(scopes) {
        await this._config.authProvider.getAccessToken(scopes);
    }
    /**
     * Retrieves information about your access token.
     */
    async getTokenInfo() {
        try {
            const data = await this.callApi({ type: 'auth', url: 'validate' });
            return new auth_1.TokenInfo(data);
        }
        catch (e) {
            if (e instanceof api_call_1.HttpStatusCodeError && e.statusCode === 401) {
                throw new auth_1.InvalidTokenError();
            }
            throw e;
        }
    }
    /**
     * Makes a call to the Twitch API using your access token.
     *
     * @param options The configuration of the call.
     */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    async callApi(options) {
        var _a;
        const { authProvider } = this._config;
        const shouldAuth = (_a = options.auth) !== null && _a !== void 0 ? _a : true;
        let accessToken = shouldAuth
            ? await authProvider.getAccessToken(options.scope ? [options.scope] : undefined)
            : null;
        if (!accessToken) {
            return await api_call_1.callTwitchApi(options, authProvider.clientId, undefined, this._config.fetchOptions);
        }
        if (auth_1.accessTokenIsExpired(accessToken) && authProvider.refresh) {
            const newAccessToken = await authProvider.refresh();
            if (newAccessToken) {
                accessToken = newAccessToken;
            }
        }
        let response = await this._callApiInternal(options, authProvider.clientId, accessToken.accessToken);
        if (response.status === 401 && authProvider.refresh) {
            await authProvider.refresh();
            accessToken = await authProvider.getAccessToken(options.scope ? [options.scope] : []);
            if (accessToken) {
                response = await this._callApiInternal(options, authProvider.clientId, accessToken.accessToken);
            }
        }
        return await api_call_1.transformTwitchApiResponse(response);
    }
    /**
     * A group of Kraken API methods.
     *
     * @deprecated Use Helix wherever possible.
     */
    get kraken() {
        return new KrakenApiGroup_1.KrakenApiGroup(this);
    }
    /**
     * A group of Helix API methods.
     */
    get helix() {
        return new HelixApiGroup_1.HelixApiGroup(this);
    }
    /**
     * The API methods that deal with badges.
     */
    get badges() {
        return new BadgesApi_1.BadgesApi(this);
    }
    /**
     * Various API methods that are not officially supported by Twitch.
     */
    get unsupported() {
        return new UnsupportedApi_1.UnsupportedApi(this);
    }
    /** @private */
    get _authProvider() {
        return this._config.authProvider;
    }
    async _callApiInternal(options, clientId, accessToken) {
        const { fetchOptions } = this._config;
        if (options.type === 'helix') {
            return await this._helixRateLimiter.request({ options, clientId, accessToken, fetchOptions });
        }
        return await api_call_1.callTwitchApiRaw(options, clientId, accessToken, fetchOptions);
    }
};
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], ApiClient.prototype, "kraken", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], ApiClient.prototype, "helix", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], ApiClient.prototype, "badges", null);
tslib_1.__decorate([
    cache_decorators_1.CachedGetter()
], ApiClient.prototype, "unsupported", null);
ApiClient = tslib_1.__decorate([
    cache_decorators_1.Cacheable,
    common_1.rtfm('api', 'ApiClient')
], ApiClient);
exports.ApiClient = ApiClient;
