// @twurple/api/lib/api/helix/channel/HelixChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixChannel = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A Twitch channel.
 */
let HelixChannel = class HelixChannel {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The ID of the channel.
     */
    get id() {
        return this._data.broadcaster_id;
    }
    /**
     * The name of the channel.
     */
    get name() {
        return this._data.broadcaster_login;
    }
    /**
     * The display name of the channel.
     */
    get displayName() {
        return this._data.broadcaster_name;
    }
    /**
     * The language of the channel.
     */
    get language() {
        return this._data.broadcaster_language;
    }
    /**
     * The ID of the game currently played on the channel.
     */
    get gameId() {
        return this._data.game_id;
    }
    /**
     * The name of the game currently played on the channel.
     */
    get gameName() {
        return this._data.game_name;
    }
    /**
     * Retrieves information about the game that is being played on the stream.
     */
    async getGame() {
        return (await this._client.helix.games.getGameById(this._data.game_id));
    }
    /**
     * The title of the channel.
     */
    get title() {
        return this._data.title;
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixChannel.prototype, "_data", void 0);
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixChannel.prototype, "_client", void 0);
HelixChannel = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixChannel', 'id')
], HelixChannel);
exports.HelixChannel = HelixChannel;
