// @twurple/api/lib/api/helix/moderation/HelixBanEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixBanEvent = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
const HelixBan_1 = require("./HelixBan");
/**
 * An event that indicates the change of a ban status, i.e. banning or unbanning a user.
 *
 * @inheritDoc
 */
let HelixBanEvent = class HelixBanEvent extends HelixBan_1.HelixBan {
    /** @private */
    constructor(eventData, client) {
        super(eventData.event_data, client);
        this._eventData = eventData;
    }
    /**
     * The unique ID of the ban event.
     */
    get eventId() {
        return this._eventData.id;
    }
    /**
     * The type of the ban event.
     */
    get eventType() {
        return this._eventData.event_type;
    }
    /**
     * The date of the ban event.
     */
    get eventDate() {
        return new Date(this._eventData.event_timestamp);
    }
    /**
     * The version of the ban event.
     */
    get eventVersion() {
        return this._eventData.version;
    }
    /**
     * The id of the broadcaster from whose chat the user was banned/unbanned.
     */
    get broadcasterId() {
        return this._eventData.event_data.broadcaster_id;
    }
    /**
     * The name of the broadcaster from whose chat the user was banned/unbanned.
     */
    get broadcasterName() {
        return this._eventData.event_data.broadcaster_login;
    }
    /**
     * The display name of the broadcaster from whose chat the user was banned/unbanned.
     */
    get broadcasterDisplayName() {
        return this._eventData.event_data.broadcaster_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._eventData.event_data.broadcaster_id));
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixBanEvent.prototype, "_eventData", void 0);
HelixBanEvent = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixBanEvent', 'userId')
], HelixBanEvent);
exports.HelixBanEvent = HelixBanEvent;
