// @twurple/api/lib/api/helix/moderation/HelixModeratorEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixModeratorEvent = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
const HelixModerator_1 = require("./HelixModerator");
/**
 * An event that indicates the change of a moderator status, i.e. gaining or losing moderation privileges.
 */
let HelixModeratorEvent = class HelixModeratorEvent extends HelixModerator_1.HelixModerator {
    /** @private */
    constructor(eventData, client) {
        super(eventData.event_data, client);
        this._eventData = eventData;
    }
    /**
     * The unique ID of the moderator event.
     */
    get eventId() {
        return this._eventData.id;
    }
    /**
     * The type of the moderator event.
     */
    get eventType() {
        return this._eventData.event_type;
    }
    /**
     * The date of the moderator event.
     */
    get eventDate() {
        return new Date(this._eventData.event_timestamp);
    }
    /**
     * The version of the moderator event.
     */
    get eventVersion() {
        return this._eventData.version;
    }
    /**
     * The id of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterId() {
        return this._eventData.event_data.broadcaster_id;
    }
    /**
     * The name of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterName() {
        return this._eventData.event_data.broadcaster_login;
    }
    /**
     * The display name of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterDisplayName() {
        return this._eventData.event_data.broadcaster_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.helix.users.getUserById(this._eventData.event_data.broadcaster_id));
    }
};
HelixModeratorEvent = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixModeratorEvent', 'userId')
], HelixModeratorEvent);
exports.HelixModeratorEvent = HelixModeratorEvent;
