// @twurple/api/lib/api/kraken/channel/PrivilegedChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivilegedChannel = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
const Channel_1 = require("./Channel");
/**
 * A channel you have extended privileges for, i.e. the channel of the currently authenticated user.
 */
let PrivilegedChannel = class PrivilegedChannel extends Channel_1.Channel {
    /**
     * The channel's stream key.
     */
    get streamKey() {
        return this._data.stream_key;
    }
    /**
     * The channel's email address.
     */
    get email() {
        return this._data.email;
    }
    /**
     * Retrieves the list of editors of the channel.
     */
    async getEditors() {
        return await this._client.kraken.channels.getChannelEditors(this);
    }
    /**
     * Starts a commercial in the channel.
     *
     * @param length The length of the commercial.
     */
    async startCommercial(length) {
        await this._client.kraken.channels.startChannelCommercial(this, length);
    }
    /**
     * Resets the given channel's stream key.
     */
    async resetStreamKey() {
        const channelData = await this._client.kraken.channels.resetChannelStreamKey(this);
        const streamKey = channelData.streamKey;
        this._data.stream_key = streamKey;
        return streamKey;
    }
};
PrivilegedChannel = tslib_1.__decorate([
    common_1.rtfm('api', 'PrivilegedChannel', 'id')
], PrivilegedChannel);
exports.PrivilegedChannel = PrivilegedChannel;
