// @twurple/api/lib/api/kraken/team/Team.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Team = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A Twitch team.
 */
let Team = class Team {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The ID of the team.
     */
    get id() {
        return this._data._id;
    }
    /**
     * The background url of the team.
     */
    get background() {
        return this._data.background;
    }
    /**
     * The banner url of the team.
     */
    get banner() {
        return this._data.banner;
    }
    /**
     * The date when the team was created.
     */
    get creationDate() {
        return new Date(this._data.created_at);
    }
    /**
     * The last date when the team changed anything.
     */
    get updateDate() {
        return new Date(this._data.updated_at);
    }
    /**
     * The name of the team.
     */
    get name() {
        return this._data.name;
    }
    /**
     * The info of the team.
     */
    get info() {
        return this._data.info;
    }
    /**
     * The display name of the team.
     */
    get displayName() {
        return this._data.display_name;
    }
    /**
     * The URL to the profile picture of the team.
     */
    get logoUrl() {
        return this._data.logo;
    }
    async getUsers() {
        const team = await this._client.kraken.teams.getTeamByName(this.name);
        return await team.getUsers();
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], Team.prototype, "_data", void 0);
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], Team.prototype, "_client", void 0);
Team = tslib_1.__decorate([
    common_1.rtfm('api', 'Team', 'id')
], Team);
exports.Team = Team;
