// @twurple/api/es/api/kraken/channel/PrivilegedChannel.mjs
import { __decorate } from "tslib";
import { rtfm } from '@twurple/common';
import { Channel } from "./Channel.mjs";
/**
 * A channel you have extended privileges for, i.e. the channel of the currently authenticated user.
 */
let PrivilegedChannel = class PrivilegedChannel extends Channel {
    /**
     * The channel's stream key.
     */
    get streamKey() {
        return this._data.stream_key;
    }
    /**
     * The channel's email address.
     */
    get email() {
        return this._data.email;
    }
    /**
     * Retrieves the list of editors of the channel.
     */
    async getEditors() {
        return await this._client.kraken.channels.getChannelEditors(this);
    }
    /**
     * Starts a commercial in the channel.
     *
     * @param length The length of the commercial.
     */
    async startCommercial(length) {
        await this._client.kraken.channels.startChannelCommercial(this, length);
    }
    /**
     * Resets the given channel's stream key.
     */
    async resetStreamKey() {
        const channelData = await this._client.kraken.channels.resetChannelStreamKey(this);
        const streamKey = channelData.streamKey;
        this._data.stream_key = streamKey;
        return streamKey;
    }
};
PrivilegedChannel = __decorate([
    rtfm('api', 'PrivilegedChannel', 'id')
], PrivilegedChannel);
export { PrivilegedChannel };
