// @twurple/api/lib/api/helix/stream/HelixStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixStream = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A Twitch stream.
 */
let HelixStream = class HelixStream {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The stream ID.
     */
    get id() {
        return this._data.id;
    }
    /**
     * The user ID.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The user's name.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The user's display name.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * Retrieves information about the user broadcasting the stream.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
    /**
     * The game ID.
     */
    get gameId() {
        return this._data.game_id;
    }
    /**
     * The game name.
     */
    get gameName() {
        return this._data.game_name;
    }
    /**
     * Retrieves information about the game that is being played on the stream.
     */
    async getGame() {
        return (await this._client.helix.games.getGameById(this._data.game_id));
    }
    /**
     * The type of the stream.
     */
    get type() {
        return this._data.type;
    }
    /**
     * The title of the stream.
     */
    get title() {
        return this._data.title;
    }
    /**
     * The number of viewers the stream currently has.
     */
    get viewers() {
        return this._data.viewer_count;
    }
    /**
     * The time when the stream started.
     */
    get startDate() {
        return new Date(this._data.started_at);
    }
    /**
     * The language of the stream.
     */
    get language() {
        return this._data.language;
    }
    /**
     * The URL of the thumbnail of the stream.
     */
    get thumbnailUrl() {
        return this._data.thumbnail_url;
    }
    /**
     * Builds the thumbnail URL of the stream using the given dimensions.
     *
     * @param width The width of the thumbnail.
     * @param height The height of the thumbnail.
     */
    getThumbnailUrl(width, height) {
        return this._data.thumbnail_url.replace('{width}', width.toString()).replace('{height}', height.toString());
    }
    /**
     * The IDs of the tags of the stream.
     */
    get tagIds() {
        var _a;
        return (_a = this._data.tag_ids) !== null && _a !== void 0 ? _a : [];
    }
    /**
     * Retrieves the tags of the stream.
     */
    async getTags() {
        var _a;
        return await this._client.helix.tags.getStreamTagsByIds((_a = this._data.tag_ids) !== null && _a !== void 0 ? _a : []);
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixStream.prototype, "_data", void 0);
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixStream.prototype, "_client", void 0);
HelixStream = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixStream', 'id')
], HelixStream);
exports.HelixStream = HelixStream;
