// @twurple/api/lib/api/helix/bits/HelixBitsLeaderboardEntry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixBitsLeaderboardEntry = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * A Bits leaderboard entry.
 */
let HelixBitsLeaderboardEntry = class HelixBitsLeaderboardEntry {
    /** @private */
    constructor(data, client) {
        this._data = data;
        this._client = client;
    }
    /**
     * The ID of the user on the leaderboard.
     */
    get userId() {
        return this._data.user_id;
    }
    /**
     * The name of the user on the leaderboard.
     */
    get userName() {
        return this._data.user_login;
    }
    /**
     * The display name of the user on the leaderboard.
     */
    get userDisplayName() {
        return this._data.user_name;
    }
    /**
     * The position of the user on the leaderboard.
     */
    get rank() {
        return this._data.rank;
    }
    /**
     * The amount of bits used in the given period of time.
     */
    get amount() {
        return this._data.score;
    }
    /**
     * Retrieves the user of entry on the leaderboard.
     */
    async getUser() {
        return (await this._client.helix.users.getUserById(this._data.user_id));
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixBitsLeaderboardEntry.prototype, "_data", void 0);
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixBitsLeaderboardEntry.prototype, "_client", void 0);
HelixBitsLeaderboardEntry = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixBitsLeaderboardEntry', 'userId')
], HelixBitsLeaderboardEntry);
exports.HelixBitsLeaderboardEntry = HelixBitsLeaderboardEntry;
